### PR TITLE
Initial version of batch requests.

### PR DESCRIPTION
**Pull request by tomusdrw**
(Still in progress (need to get rid of Box), but would like you guys to have a look already).

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -104,8 +140,23 @@ impl<X, T> Transport for X where
...
+}
+
+impl<X, T> BatchTransport for X where
+  T: BatchTransport + ?Sized,
+  X: ::std::ops::Deref<Target=T>,
+{
+  type Batch = T::Batch;
+
+  fn send_batch(&self, requests: Vec<(RequestId, rpc::Call)>) -> Self::Batch {
+    (**self).send_batch(requests)
```
Comment (debris) [inline src/lib.rs]: is it really needed? I mean:

```rust
+impl<X, T> BatchTransport for X where
 +  T: BatchTransport + ?Sized,
```

File: src/transports/ipc.rs [hunk 2]
```diff
@@ -60,22 +60,24 @@ impl Drop for EventLoopHandle {
 }
 
 
-/// Error returned while initializing IPC transport
+/// Error returned while initializing IPC transport.
 pub type IpcError = io::Error;
+/// Result of initializing IPC transport.
+pub type IpcResult<T> = ::std::result::Result<T, IpcError>;
```
Comment (rphmeier) [inline src/transports/ipc.rs]: Style is typically to declare `type Result` and then it can be used as `ipc::Result`

**Comment by tomusdrw** [comment]
Ping: @debris @rphmeier

File: src/transports/batch.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+use transports::Result;
+use {BatchTransport, Transport, Error as RpcError, RequestId};
+
+type Pending = oneshot::Sender<Result<rpc::Value>>;
+
+/// Transport allowing to batch queries together.
+pub struct Batch<T> {
+  transport: T,
+  pending: Arc<Mutex<BTreeMap<RequestId, Pending>>>,
+  batch: Mutex<Vec<(RequestId, rpc::Call)>>,
```
Comment (rphmeier) [inline src/transports/batch.rs]: could probably kill `pending` field and put it inside the mutex here as far as I can tell each batch should have its own pending set.

Comment (rphmeier) [inline src/transports/batch.rs]: i guess it's a slight optimization that you don't have to reallocate the pending set, but it also means that queuing a new batch will be blocked while processing the response of the last one.

File: src/transports/ipc.rs [hunk 4]
```diff
@@ -156,44 +157,91 @@ impl Ipc {
...
+    let (tx, rx) = futures::oneshot();
+    self.pending.lock().insert(id, tx);
+    let result = {
+      let mut sender = self.write_sender.lock();
+      (*sender).send(request.into_bytes())
+    };
 
     IpcTask {
       state: IpcTaskState::Sending(Some(result), rx),
+      extract: batch_response as fn(Vec<RpcResult<rpc::Value>>) -> RpcResult<Vec<RpcResult<rpc::Value>>>,
```
Comment (rphmeier) [inline src/transports/ipc.rs]: you don't technically need a separate `batch_response` function for this: enum variants are their own functions so I think `Ok as fn(...) -> ...` would work

**Comment by rphmeier** [comment]
Passing a `Vec<(RequestId, Call)>` to `send_batch` when the implementations just take out the calls (and allocate again) doesn't seem quite right. Passing an `I: IntoIterator<Item=(RequestId, Call)>` would be a lot more flexible. But I guess it would bar any future implementation of `BatchTransport` for type-erased transport until specialization is stable.

Also a utility method on `Transport`: `fn batched(self) -> Batch<Self> where Self: Sized + BatchTransport` would be cool

**Comment by tomusdrw** [comment]
Updated.
- [x] `send_batch` now takes `IntoIterator` instead of `Vec`
- [x] `submit_batch` returns concrete type instead of `BoxFuture`.

**Pull request merged**